grafana/resource_machine_learning_outlier.go

```diff
@@ -0,0 +1,268 @@
+package grafana
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	"github.com/grafana/machine-learning-go-client/mlapi"
+)
+
+func ResourceMachineLearningOutlierDetector() *schema.Resource {
+	return &schema.Resource{
+
+		Description: `
+An outlier detector monitors the results of a query and reports when its values are outside normal bands.
+
+The normal band is configured by choice of algorithm, its sensitivity and other configuration.
+
+Visit https://grafana.com/docs/grafana-cloud/machine-learning/outlier-detection/ for mroe details.
+`,
+
+		CreateContext: resourceMachineLearningOutlierCreate,
+		ReadContext:   resourceMachineLearningOutlierRead,
+		UpdateContext: resourceMachineLearningOutlierUpdate,
+		DeleteContext: resourceMachineLearningOutlierDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Description: "The ID of the outlier detector.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"name": {
+				Description: "The name of the outlier detector.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"metric": {
+				Description: "The metric used to query the outlier detector results.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"description": {
+				Description: "A description of the outlier detector.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"datasource_id": {
+				Description: "The id of the datasource to query.",
+				Type:        schema.TypeInt,
+				Optional:    true,
+			},
+			"datasource_uid": {
+				Description: "The uid of the datasource to query.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"datasource_type": {
+				Description: "The type of datasource being queried. Currently allowed values are prometheus, graphite, loki, postgres, and datadog.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"query_params": {
+				Description: "An object representing the query params to query Grafana with.",
+				Type:        schema.TypeMap,
+				Required:    true,
+			},
+			"interval": {
+				Description: "The data interval in seconds to monitor.",
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     300,
+			},
+			"algorithm": {
+				Description: "The algorithm to use and its configuration. See https://grafana.com/docs/grafana-cloud/machine-learning/outlier-detection/ for details.",
+				Type:        schema.TypeSet,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type: schema.TypeString,
+							// ExactlyOneOf: []string{"mad", "dbscan"},
```

Comment:
Is this commented out in purpose?

Reply:
I used a validator instead, as you suggested above